File: app/src/main/java/com/haivo/diana/View/Button.java
package com.haivo.diana.View;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import com.haivo.diana.R;

/**
 * TODO: document your custom view class.
 */
public class Button extends LinearLayout {
    private String buttonText;
    private String buttonVariant;
    private Boolean buttonGhost;

    private View rootView;
    private TextView textView;

    public Button(Context context) {
        super(context);
        init(null, 0);
    }

    public Button(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public Button(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        // Load attributes
        final TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.Button, defStyle, 0);

        buttonText = a.getString(R.styleable.Button_text);

        buttonVariant = a.getString(R.styleable.Button_variant);

        buttonGhost = a.getBoolean(R.styleable.Button_ghost, false);

        a.recycle();

        rootView = inflate(getContext(), R.layout.view_button, this);
        textView = rootView.findViewById(R.id.button);
        textView.setText(buttonText);
        textView.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                rootView.onTouchEvent(motionEvent);
                return false;
            }
        });

        rootView.setPadding(getPaddingLeft(), getPaddingTop(), getPaddingRight(), getPaddingBottom());

        updateStyle();
    }

    public void updateStyle() {
        if (!buttonGhost) {
            if (buttonVariant.equals(getContext().getString(R.string.button_variant_white))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_white));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorWhite));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_accent))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_accent));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorAccent));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_danger))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_danger));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorDanger));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_fade))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_white));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorTextSecondary));
            }
        } else {
            if (buttonVariant.equals(getContext().getString(R.string.button_variant_white))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_ghost_white));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorWhite));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_accent))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_ghost_accent));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorAccent));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_danger))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_ghost_danger));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorDanger));
            } else if (buttonVariant.equals(getContext().getString(R.string.button_variant_fade))) {
                rootView.setBackground(getContext().getDrawable(R.drawable.button_bg_ghost_white));
                textView.setTextColor(ContextCompat.getColor(getContext(), R.color.colorTextSecondary));
            }
        }
    }

    public void setText(String text) {
        textView.setText(text);
    }
}

File: app/src/main/java/com/haivo/diana/View/Button.kt
package com.haivo.diana.View

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.haivo.diana.R

/**
 * TODO: document your custom view class.
 */
class Button : LinearLayout {
    private var buttonText: String?
    private var buttonVariant: String?
    private var buttonGhost: Boolean

    private val buttonView: View
    private val textView: TextView

    constructor(context: Context) :
            this(context, null)

    constructor(context: Context, attrs: AttributeSet?) :
            this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr) {
        // Load attributes
        val a = context.obtainStyledAttributes(attrs, R.styleable.Button, defStyleAttr, 0)

        buttonText = a.getString(R.styleable.Button_text)

        buttonVariant = a.getString(R.styleable.Button_variant)

        buttonGhost = a.getBoolean(R.styleable.Button_ghost, false)

        a.recycle()

        buttonView = View.inflate(context, R.layout.view_button, this)
        textView = buttonView.findViewById(R.id.button)
        textView.text = buttonText
        textView.setOnTouchListener { _, motionEvent ->
            buttonView.onTouchEvent(motionEvent)
            false
        }

        buttonView.setPadding(paddingLeft, paddingTop, paddingRight, paddingBottom)

        updateStyle()
    }

    fun updateStyle() {
        if (!buttonGhost) {
            when (buttonVariant) {
                context.getString(R.string.button_variant_white) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_white)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorWhite))
                }
                context.getString(R.string.button_variant_accent) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_accent)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorAccent))
                }
                context.getString(R.string.button_variant_danger) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_danger)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorDanger))
                }
                context.getString(R.string.button_variant_fade) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_white)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorTextSecondary))
                }
            }
        } else {
            when (buttonVariant) {
                context.getString(R.string.button_variant_white) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_ghost_white)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorWhite))
                }
                context.getString(R.string.button_variant_accent) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_ghost_accent)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorAccent))
                }
                context.getString(R.string.button_variant_danger) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_ghost_danger)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorDanger))
                }
                context.getString(R.string.button_variant_fade) -> {
                    buttonView.background = context.getDrawable(R.drawable.button_bg_ghost_white)
                    textView.setTextColor(ContextCompat.getColor(context, R.color.colorTextSecondary))
                }
            }
        }
    }

    fun setText(text: String) {
        textView.text = text
    }
}
